### PR TITLE
fix UNION ALL bug: thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1', ./src/datatypes/schema.rs:165:10

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1064

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/src/optimizer/projection_push_down.rs [hunk 1]
```diff
@@ -417,9 +416,15 @@ fn optimize_plan(
                 })
                 .collect::<Result<Vec<_>>>()?;
 
+            let mut new_schema = Vec::new();
+            for df_field in schema.fields() {
+                if union_required_fields.contains(df_field.field()) {
+                    new_schema.push(df_field.clone());
+                }
+            }
```
Comment (alamb) [inline datafusion/src/optimizer/projection_push_down.rs:424]: I don't know if you are into the whole functional programming thing, but you can probably write this without a `mut Vec` like

```rust
            let new_schema = DFSchema::new(
                schema.fields()
                    .iter()
                    .filter(|df_field| {
                        union_required_fields.contains(df_field.field())
                    })
                    .cloned()
                    .collect()
            )?;
```

I don't think it makes any practical difference (what you have here is fine too) I just figured I would point it out

Comment (houqp) [reply datafusion/src/optimizer/projection_push_down.rs:424]: I would prefer us to go functional ere to keep the style consistent within the code base :)

**Review by houqp — APPROVED** [review]
Good catch @xudong963 !

File: datafusion/src/optimizer/projection_push_down.rs [hunk 2]
```diff
@@ -417,9 +416,15 @@ fn optimize_plan(
                 })
                 .collect::<Result<Vec<_>>>()?;
 
+            let mut new_schema = Vec::new();
+            for df_field in schema.fields() {
+                if union_required_fields.contains(df_field.field()) {
```
Comment (Dandandan) [inline datafusion/src/optimizer/projection_push_down.rs:421]: This has quadratic complexity, which can be problematic when we have many fields but I think we have more areas in the code base with this issue, so not really needed to solve now.

Comment (houqp) [reply datafusion/src/optimizer/projection_push_down.rs:421]: given that `union_required_fields` is a hashset, shouldn't the cost be close to constant?

**Pull request merged**